File: backend-python/app/schemas/customer.py
import re
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.schemas.customer_contact import (
    CustomerContactCreate,
    CustomerContactResponse,
    CustomerContactUpdate,
)


class CustomerBase(BaseModel):
    name: str = Field(..., max_length=100, description="客户单位")
    address: str | None = Field(None, max_length=200, description="客户地址")
    remarks: str | None = Field(None, description="备注")


class CustomerCreate(CustomerBase):
    contacts: list[CustomerContactCreate] | None = Field(None, description="联系人列表")

    @field_validator('contacts', mode='before')
    @classmethod
    def validate_contacts(cls, v):
        if v is None:
            return []
        return v


class CustomerUpdate(BaseModel):
    name: str | None = Field(None, max_length=100, description="客户单位")
    address: str | None = Field(None, max_length=200, description="客户地址")
    remarks: str | None = Field(None, description="备注")
    contacts: list[CustomerContactCreate] | None = Field(None, description="联系人列表（完整替换）")


class CustomerResponse(CustomerBase):
    id: int
    contact_person: str | None = None
    phone: str | None = None
    contact_position: str | None = None
    contacts: list[CustomerContactResponse] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class CustomerListResponse(BaseModel):
    content: list[CustomerResponse]
    totalElements: int
    totalPages: int
    size: int
    number: int

File: src/services/customer.ts
import request from '../api/request'
import { API_ENDPOINTS } from '../api/endpoints'
import type { ApiResponse, Customer as BaseCustomer } from '@sstcp/shared'

export interface CustomerContact {
  id: number
  customer_id: number
  contact_person: string
  phone: string | null
  contact_position: string | null
  remarks: string | null
  created_at: string
  updated_at: string
}

export interface CustomerContactCreate {
  contact_person: string
  phone?: string
  contact_position?: string
  remarks?: string
}

export interface Customer extends BaseCustomer {
  contacts: CustomerContact[]
}

export interface CustomerCreate {
  name: string
  address?: string
  contacts?: CustomerContactCreate[]
  remarks?: string
}

export interface CustomerUpdate {
  name?: string
  address?: string
  contacts?: CustomerContactCreate[]
  remarks?: string
}

export interface CustomerListParams {
  page?: number
  size?: number
  name?: string
  contact_person?: string
}

export interface CustomerListResponse {
  content: Customer[]
  totalElements: number
  totalPages: number
  size: number
  number: number
}

export const customerService = {
  async getList(
    params: CustomerListParams,
    signal?: AbortSignal
  ): Promise<ApiResponse<CustomerListResponse>> {
    const response = await request.get(API_ENDPOINTS.CUSTOMER.LIST, { params, signal })
    return response as unknown as ApiResponse<CustomerListResponse>
  },

  async getById(id: number, signal?: AbortSignal): Promise<ApiResponse<Customer>> {
    const response = await request.get(API_ENDPOINTS.CUSTOMER.DETAIL(id), { signal })
    return response as unknown as ApiResponse<Customer>
  },

  async create(data: CustomerCreate): Promise<ApiResponse<Customer>> {
    const response = await request.post(API_ENDPOINTS.CUSTOMER.LIST, data)
    return response as unknown as ApiResponse<Customer>
  },

  async update(id: number, data: CustomerUpdate): Promise<ApiResponse<Customer>> {
    const response = await request.put(API_ENDPOINTS.CUSTOMER.DETAIL(id), data)
    return response as unknown as ApiResponse<Customer>
  },

  async delete(id: number, cascade: boolean = false): Promise<ApiResponse<void>> {
    const response = await request.delete(API_ENDPOINTS.CUSTOMER.DETAIL(id), {
      params: { cascade },
    })
    return response as unknown as ApiResponse<void>
  },
}
